File: app/src/main/java/projects/csce/evence/view/ui/GenerateQR.java
package projects.csce.evence.view.ui;

import android.content.Intent;
import android.graphics.Bitmap;
import android.net.Uri;
import android.os.Bundle;
import android.os.ParcelFileDescriptor;
import android.util.Log;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.databinding.DataBindingUtil;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProvider;
import androidx.lifecycle.ViewModelProviders;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.util.Collections;
import java.util.Objects;

import javax.inject.Inject;

import daniel.perez.ical.EventSpec;
import daniel.perez.ical.ICalSpec;
import daniel.perez.ical.Parser;
import okio.BufferedSink;
import okio.Okio;
import projects.csce.evence.R;
import projects.csce.evence.databinding.ActivityGenerateQrBinding;
import projects.csce.evence.service.model.FileManager;
import projects.csce.evence.service.model.qr.QrAttempt;
import projects.csce.evence.service.model.qr.QrBitmapGenerator;
import projects.csce.evence.utils.Utils;
import projects.csce.evence.view.ui.model.ViewCalendarData;
import projects.csce.evence.view.ui.model.ViewEvent;
import projects.csce.evence.viewmodel.GenerateQrViewModel;

public class GenerateQR extends AppCompatActivity implements Observer<QrAttempt> {
    private GenerateQrViewModel viewModel;
    private ActivityGenerateQrBinding binding;
    private ICalSpec currentEvent;
    @Inject ViewModelProvider.Factory viewModelFactory;
    @Inject QrBitmapGenerator generator;
    @Inject FileManager fileManager;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        Utils.getAppComponent(this).inject(this);
        super.onCreate(savedInstanceState);
        binding = DataBindingUtil.setContentView(this, R.layout.activity_generate_qr);
        viewModel = ViewModelProviders.of(this, viewModelFactory).get(GenerateQrViewModel.class);
        binding.setView(this);
        binding.setViewModel(viewModel);

        viewModel.qrImages().observe(this, this);

        if(getIntent() != null && getIntent().getStringExtra("FILE_PATH") != null)
        {
            File file = new File(getIntent().getStringExtra("FILE_PATH"));
            ICalSpec ical = Parser.INSTANCE.parse(file);
            currentEvent = ical;
            Log.i("ICALFILEPATH", ical.getEvents().get(0).toString());
            fillInFields();
        }
    }

    //fill in editText and other fields when event is being edited
    public void fillInFields(){
        EventSpec event = currentEvent.getEvents().get(0);
        binding.titleEditText.setText(event.getTitle());
        binding.startDateTextView.setText(event.getStartDate());
        binding.startTimeTextView.setText(event.getStartTime());
        binding.endDateTextView.setText(event.getEndFormatted("MM-dd-yyyy"));
        binding.endTimeTextView.setText(event.getEndFormatted("hh:mm a"));
        binding.locationEditText.setText(event.getLocation());
        binding.descriptionEditText.setText(event.getDescription());


    }



    public void onChanged(QrAttempt attempt) {
        if (attempt instanceof QrAttempt.Success) {
            Bitmap bitmap = ((QrAttempt.Success) attempt).getBitmap();
            //binding.QRImage.setImageBitmap(bitmap);
        } else if (attempt instanceof QrAttempt.Failure) {
            Throwable e = ((QrAttempt.Failure) attempt).getE();
            e.printStackTrace();
            Utils.toastShort(getApplicationContext(), Objects.requireNonNull(e.getLocalizedMessage()));
        }
    }

    public void generateQR() {
        // Handle the dates
        int[] startMonthDayYear = Utils.toInts(binding.startDateTextView.getText().toString().split("/"));
        int[] endMonthDayYear = Utils.toInts(binding.endDateTextView.getText().toString().split("/"));

        // Handle the hours and minutes
        int[] startHourMinutes = Utils.toInts(binding.startTimeTextView.getText().toString().split(":"));
        int[] endHourMinutes = Utils.toInts(binding.endTimeTextView.getText().toString().split(":"));



        EventSpec event = new EventSpec.Builder(0)
                .title(binding.titleEditText.getText().toString())
                .description(binding.descriptionEditText.getText().toString())
                .location(binding.locationEditText.getText().toString())
                .start(Utils.toZonedDateTime(startMonthDayYear, startHourMinutes))
                .end(Utils.toZonedDateTime(endMonthDayYear, endHourMinutes))
                .build();

        currentEvent = new ICalSpec.Builder()
                .fileName(binding.titleEditText.getText().toString())
                .addEvent(event)
                .build();

        // Write the file to the file system
        viewModel.saveFile(currentEvent);

        ViewEvent viewEvent = new ViewEvent(event.getTitle(),
                event.getDescription(),
                event.getStartTime(),
                event.getStartDate(),
                event.getStartInstantEpoch(),
                event.getEndEpochMilli(),
                event.getLocation(),
                event.text(),
                generator.forceGenerate(event.text())
        );
        ViewCalendarData calendar = new ViewCalendarData(currentEvent.getFileName(), Collections.singletonList(viewEvent));
        QRDialog qrDialog = new QRDialog(this, calendar, fileManager);
    }

    // Handle the user choosing a place to store the file
    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == 1) {
            Uri uri;
            if (data != null) {
                uri = data.getData();
                try {
                    ParcelFileDescriptor pfd = getContentResolver().openFileDescriptor(uri, "w");

                    FileOutputStream fileOutputStream = new FileOutputStream(pfd.getFileDescriptor());
                    try (BufferedSink sink = Okio.buffer(Okio.sink(fileOutputStream))) {
                        sink.writeUtf8(currentEvent.text());
                    } catch (IOException e) {
                        Log.i("FILEWRITER", e.getMessage());
                        e.printStackTrace();
                    }
                    pfd.close();
                } catch (FileNotFoundException e) {
                    Log.i("FILEWRITER", e.getMessage());
                    e.printStackTrace();
                } catch (IOException e) {
                    e.printStackTrace();
                }
                Log.i("FILEWRITER", "Wrote File");
            }
        }
    }

    public void startDateDialog() {
        CalendarDialog calendarDialog = new CalendarDialog(this, binding.startDateTextView);
        calendarDialog.dateDialog();
    }

    public void startTimeDialog() {
        CalendarDialog calendarDialog = new CalendarDialog(this, binding.startTimeTextView);
        calendarDialog.timeDialog();
    }

    public void endDateDialog() {
        CalendarDialog calendarDialog = new CalendarDialog(this, binding.endDateTextView);
        calendarDialog.dateDialog();
    }

    public void endTimeDialog() {
        CalendarDialog calendarDialog = new CalendarDialog(this, binding.endTimeTextView);
        calendarDialog.timeDialog();
    }



}

File: GenerateQRView/src/main/java/teamevence/evenceapp/generateqrview/GenerateQR.kt
package teamevence.evenceapp.generateqrview

import android.annotation.SuppressLint
import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.AdapterView
import android.widget.Toast
import androidx.lifecycle.ViewModelProvider
import com.jakewharton.rxbinding4.view.clicks
import dagger.hilt.android.AndroidEntryPoint
import teamevence.evenceapp.core.*
import teamevence.evenceapp.core.db.UiNewEvent
import teamevence.evenceapp.core.db.toViewEvent
import teamevence.evenceapp.core.model.DateSetEvent
import teamevence.evenceapp.core.model.Half
import teamevence.evenceapp.core.model.TimeSetEvent
import teamevence.evenceapp.core.model.ViewEvent
import teamevence.evenceapp.generateqrview.databinding.ActivityGenerateQrBinding
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers
import timber.log.Timber
import java.time.LocalDateTime
import javax.inject.Inject

@AndroidEntryPoint
class GenerateQR : BaseActivity(), DialogClosable, AdapterView.OnItemSelectedListener {
    private lateinit var viewModel: GenerateQrViewModel
    private lateinit var binding: ActivityGenerateQrBinding
    private lateinit var currentEvent: ViewEvent
    private var startTime = TimeSetEvent(0, 0, Half.AM)
    private var endTime = TimeSetEvent(0, 0, Half.AM)
    private var startDate = DateSetEvent(1, 31, 1999)
    private var endDate = DateSetEvent(1, 31, 1999)
    @Inject
    lateinit var dialogStarter: DialogStarter
    @Inject
    lateinit var activityResultActions: ActivityResultActions
    private var isEditing = false

    @SuppressLint("CheckResult")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityGenerateQrBinding.inflate(layoutInflater)
        setContentView(binding.root)
        viewModel = ViewModelProvider(this).get(GenerateQrViewModel::class.java)

        if (intent != null && intent.getLongExtra("EVENT_ID", -1L) != -1L) {
            isEditing = true
            val id = intent.getLongExtra("EVENT_ID", -1)
            disposables += viewModel.getEvent(id)
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe {
                        currentEvent = it
                        Timber.i(it.toString())
                        fillInFields()
                    }
        }

        /** for future updates. to set recurrence
        ArrayAdapter.createFromResource(this, R.array.event_recurrence_options_list, R.layout.spinner_text)
        .also { adapter ->
        adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
        binding.recurrenceSpinner.adapter = adapter
        }
        binding.recurrenceSpinner.onItemSelectedListener = this
         */

        disposables += binding.startDateEditText.clicks().subscribe { startDateDialog() }
        binding.startTimeEditText.clicks().subscribe { startTimeDialog() }
        binding.endDateEditText.clicks().subscribe { endDateDialog() }
        binding.endTimeEditText.clicks().subscribe { endTimeDialog() }
        binding.finishButton.clicks().subscribe {
            if (!isEditing) {
                saveEvent()
            } else {
                updateEvent()
            }
        }
    }

    //fill in editText and other fields when event is being edited
    private fun fillInFields() {
        val event = currentEvent

        //TODO("Implement the enum properly, calculate if its morning or after noon for AM, PM")
        startTime = TimeSetEvent(event.startDateTime.hour, event.startDateTime.minute, Half.AM)
        endTime = TimeSetEvent(event.endDateTime.hour, event.endDateTime.minute, Half.PM)
        startDate = DateSetEvent(event.startDateTime.monthValue, event.startDateTime.dayOfMonth, event.startDateTime.year)
        endDate = DateSetEvent(event.endDateTime.monthValue, event.endDateTime.dayOfMonth, event.endDateTime.year)

        binding.titleEditText.setText(event.title)
        binding.startDateTextView.text = event.startDatePretty()
        binding.startTimeTextView.text = event.startTimePretty()
        binding.endDateTextView.text = event.endDatePretty()
        binding.endTimeTextView.text = event.endTimePretty()
        binding.locationEditText.setText(event.location)
        binding.descriptionEditText.setText(event.description)
    }

    private fun areValidFields(): Boolean {
        if (binding.titleEditText.text.isBlank()) {
            applicationContext.snackbarShort(binding.titleEditText, getString(R.string.enter_valid_title))
            return false
        }

        if (binding.startDateTextView.text.toString().equals(getString(R.string.select_date))) {
            applicationContext.snackbarShort(binding.startDateTextView, getString(R.string.enter_valid_start_date))
            return false
        }

        //if end date is empty, set it to be the same as start date
        if (binding.endDateTextView.text.toString().equals(getString(R.string.select_date))) {
            endDate = startDate
        } else {
            if (endDate.year >= startDate.year
                    && endDate.month >= startDate.month
                    && endDate.dayOfMonth >= startDate.dayOfMonth
                    && endTime.hour >= startTime.hour
                    && endTime.minute >= startTime.minute) {
                return true
            }
            applicationContext.snackbarShort(binding.startDateTextView, getString(R.string.enter_valid_end_date_time))
            return false
        }

        return true
    }

    private fun saveEvent() {
        if (!areValidFields()) {
            return
        }
        val uiNewEvent: UiNewEvent = extractEventFromUi()
        disposables += viewModel.saveEvent(uiNewEvent)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe {
                    currentEvent = it.toViewEvent()
                    dialogStarter.startQrDialog(this, it.toViewEvent())
                }
    }

    private fun updateEvent() {
        if (!areValidFields()) {
            return
        }
        val uiNewEvent: UiNewEvent = extractEventFromUi()
        disposables += viewModel.updateEvent(currentEvent.id, uiNewEvent)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe {
                    currentEvent = it.toViewEvent()
                    dialogStarter.startQrDialog(this, it.toViewEvent())
                }
    }

    private fun extractEventFromUi(): UiNewEvent {
        Timber.i("Start_Date: ${startDate.string()}")
        Timber.i("Start_Time: $startTime")
        Timber.i("End_Date: ${endDate.string()}")
        Timber.i("End_Time: $endTime")
        val uiNewEvent: UiNewEvent
        with(binding)
        {
            uiNewEvent = UiNewEvent(
                    titleEditText.text.toString(),
                    descriptionEditText.text.toString(),
                    locationEditText.text.toString(),
                    LocalDateTime.parse("${startDate.string()}T${startTime.string()}"), //todo add method into Utils.kt
                    LocalDateTime.parse("${endDate.string()}T${endTime.string()}"),
            )
        }

        return uiNewEvent
    }

    // Handle the user choosing a place to store the file
    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == RequestCodes.REQUEST_SAF) {
            if (data != null) {
                disposables += activityResultActions.actionCreateDocumentEvent(this, currentEvent, data)
                        .subscribe {
                            when (it) {
                                ActionResult.Success -> toastShort(this,"Wrote File Successfully")
                                is ActionResult.Failure -> {
                                    Timber.e(it.t)
                                    toastShort(this, "Error writing file")
                                }
                                ActionResult.InTransit -> {
                                    Timber.i("Writing file...")
                                }
                            }
                        }
            }
        }
    }

    private fun startDateDialog() {
        disposables.add(dialogStarter.startDateDialog(this)
                .subscribe {
                    startDate = it
                    binding.startDateTextView.text = "${it.month}-${it.dayOfMonth}-${it.year}"
                })
    }

    private fun endDateDialog() {
        disposables.add(dialogStarter.startDateDialog(this)
                .subscribe {
                    endDate = it
                    binding.endDateTextView.text = "${it.month}-${it.dayOfMonth}-${it.year}"

                })
    }

    private fun startTimeDialog() {
        disposables.add(dialogStarter.startTimeDialog(this)
                .subscribe {
                    startTime = it
                    binding.startTimeTextView.text = getAMPMTimeFormat("${it.hour} ${it.minute}")
                })
    }

    private fun endTimeDialog() {
        disposables.add(dialogStarter.startTimeDialog(this)
                .subscribe {
                    endTime = it
                    binding.endTimeTextView.text = getAMPMTimeFormat("${it.hour} ${it.minute}")
                })
    }

    override fun onDestroy() {
        super.onDestroy()
        disposables.dispose()
    }

    override fun close() {
        finish()
    }

    override fun onItemSelected(parent: AdapterView<*>?, view: View?, position: Int, id: Long) {

    }

    override fun onNothingSelected(p0: AdapterView<*>?) {

    }
}